druid-shell/src/platform/x11/application.rs

```diff
@@ -0,0 +1,152 @@
+                //NOTE: I don't think we should be doing this here, but I'm trying to keep
+                //the code mostly unchanged. My personal feeling is that the best approach
+                //is to dispatch events to the window as early as possible, that is to say
+                //I would send the *raw* events to the window and then let the window figure
+                //out how to handle them. - @cmyr
+                match ev_type {
+                    xcb::EXPOSE => {
+                        let expose: &xcb::ExposeEvent = unsafe { xcb::cast_event(&ev) };
+                        let window_id = expose.window();
+                        WINDOW_MAP.with(|map| {
+                            let mut windows = map.borrow_mut();
+                            windows.get_mut(&window_id).map(|w| w.render());
+                        })
+                    }
+                    xcb::KEY_PRESS => {
+                        let key_press: &xcb::KeyPressEvent = unsafe { xcb::cast_event(&ev) };
+                        let key: u32 = key_press.detail() as u32;
+                        let key_code: KeyCode = key.into();
+
+                        let window_id = key_press.event();
+                        println!("window_id {}", window_id);
+                        WINDOW_MAP.with(|map| {
+                            let mut windows = map.borrow_mut();
+                            windows.get_mut(&window_id).map(|w| w.key_down(key_code));
+                        })
+                    }
+                    xcb::BUTTON_PRESS => {
+                        let button_press: &xcb::ButtonPressEvent = unsafe { xcb::cast_event(&ev) };
+                        let window_id = button_press.event();
+                        let mouse_event = MouseEvent {
+                            pos: Point::new(
+                                button_press.event_x() as f64,
+                                button_press.event_y() as f64,
+                            ),
+                            mods: KeyModifiers {
+                                shift: false,
+                                alt: false,
+                                ctrl: false,
+                                meta: false,
+                            },
+                            count: 0,
+                            button: MouseButton::Left,
+                        };
+                        WINDOW_MAP.with(|map| {
+                            let mut windows = map.borrow_mut();
+                            windows
+                                .get_mut(&window_id)
+                                .map(|w| w.mouse_down(&mouse_event));
+                        })
+                    }
+                    _ => {}
+                }
+            }
+        }
+    }
+
+    pub fn quit() {
+        // No-op.
+    }
+
+    pub fn clipboard() -> Clipboard {
+        // TODO(x11/clipboard): implement Application::clipboard
+        unimplemented!();
```

Comment:
for things like this I think it makes more sense to do `log::warn` with a message about what is unimplemented instead of using the unimplemented macro, which will just crash us? 🤷‍♂

Reply:
Changed *all* `unimplemented!()`s to `log::warn!()`s.

Might be overboard? Let me know if I should delete any.